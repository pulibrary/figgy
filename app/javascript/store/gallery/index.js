export const galleryState = {
  items: [],
  selected: [],
  cut: [],
  changeList: [],
  ogItems: [],
}

export const galleryMutations = {
  CUT(state, itemArray) {
    state.cut = [...itemArray]
  },
  PASTE(state, itemArray) {
    state.items = [...itemArray]
  },
  SELECT(state, itemArray) {
    state.selected = [...itemArray]
  },
  SET_GALLERY(state, items) {
    state.items = items
    state.ogItems = items
  },
  SORT_ITEMS(state, value) {
    state.items = [...value]
  },
  UPDATE_CHANGES(state, changeList) {
    state.changeList = [...changeList]
  },
  UPDATE_ITEMS(state, items) {
    state.items = [...items]
  },
}

export const galleryModule = {
  state: galleryState,
  mutations: galleryMutations,
}
