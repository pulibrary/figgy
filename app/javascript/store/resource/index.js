export const resourceState = {
  resource: {
    id: "",
    resourceClassName: "",
    bibId: "",
    label: "Resource not available.",
    thumbnail: "",
    startCanvas: "",
    isMultiVolume: false,
    viewingHint: null,
    viewingDirection: null,
    members: [],
    loadState: "NOT_LOADED",
    saveState: "NOT_SAVED",
    ogState: {},
  },
}

export const resourceMutations = {
  APPLY_STATE(state) {
    state.gallery.ogItems = state.gallery.items
    state.gallery.changeList = []
    state.resource.ogState = {
      startCanvas: resource.startPage,
      thumbnail: resource.thumbnail,
      viewingHint: resource.viewingHint,
      viewingDirection: resource.viewingDirection,
    }
    state.resource.saveState = "NOT_SAVED"
  },
  CHANGE_RESOURCE_LOAD_STATE(state, loadState) {
    state.resource.loadState = loadState
  },
  SAVED_STATE(state, saveStatus) {
    state.resource.saveState = saveStatus
  },
  SET_RESOURCE(state, resource) {
    state.resource.id = resource.id
    state.resource.resourceClassName = resource.__typename
    state.resource.label = resource.label
    state.resource.startCanvas = resource.startPage
    state.resource.viewingHint = resource.viewingHint
    state.resource.viewingDirection = resource.viewingDirection
    state.resource.thumbnail = resource.thumbnail != null ? resource.thumbnail.id : null
    state.resource.members = resource.members

    const items = resource.members.map(member => ({
      id: member.id,
      viewingHint: member.viewingHint != null ? member.viewingHint : "single",
      caption: member.label, // member.__typename + " : " + member.id,
      service:
        member["thumbnail"] && typeof(member.thumbnail.iiifServiceUrl) != "undefined"
          ? member.thumbnail.iiifServiceUrl
          : Global.figgy.resource.defaultThumbnail,
      mediaUrl:
        member["thumbnail"] && typeof(member.thumbnail.iiifServiceUrl) != "undefined"
          ? member.thumbnail.iiifServiceUrl + "/full/300,/0/default.jpg"
          : Global.figgy.resource.defaultThumbnail,
    }))
    state.gallery.items = items
    state.gallery.ogItems = items
    state.resource.loadState = "LOADED"
    state.resource.ogState = {
      startCanvas: resource.startPage,
      thumbnail: resource.thumbnail != null ? resource.thumbnail.id : null,
      viewingHint: resource.viewingHint,
      viewingDirection: resource.viewingDirection,
    }
  },
  UPDATE_STARTCANVAS(state, startCanvas) {
    state.resource.startCanvas = startCanvas
  },
  UPDATE_THUMBNAIL(state, thumbnail) {
    state.resource.thumbnail = thumbnail
  },
  UPDATE_VIEWDIR(state, viewDir) {
    state.resource.viewingDirection = viewDir
  },
  UPDATE_VIEWHINT(state, viewHint) {
    state.resource.viewingHint = viewHint
  },
}

export const resourceGetters = {
  getMemberCount: state => {
    return state.resource.members.length
  },
  isMultiVolume: state => {
    const volumes = state.resource.members.filter(member => member.__typename === "ScannedResource")
    return volumes.length > 0 ? true : false
  },
  orderChanged: state => {
    let ogOrder = JSON.stringify(state.gallery.ogItems.map(item => item.id))
    let imgOrder = JSON.stringify(state.gallery.items.map(item => item.id))
    return ogOrder !== imgOrder
  },
  stateChanged: (state, getters) => {
    var propsChanged = []
    propsChanged.push(state.resource.ogState.thumbnail !== state.resource.thumbnail)
    propsChanged.push(state.resource.ogState.startCanvas !== state.resource.startCanvas)
    propsChanged.push(state.resource.ogState.viewingHint !== state.resource.viewingHint)
    propsChanged.push(state.resource.ogState.viewingDirection !== state.resource.viewingDirection)
    propsChanged.push(state.gallery.changeList.length > 0)
    propsChanged.push(getters.orderChanged)
    if (propsChanged.indexOf(true) > -1) {
      return true
    } else {
      return false
    }
  },
}
