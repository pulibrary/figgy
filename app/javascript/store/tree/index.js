export const treeState = {
  selected: null,
}

export const treeMutations = {
  SELECT(state, selectedId) {
    state.selected = selectedId
  },
}
