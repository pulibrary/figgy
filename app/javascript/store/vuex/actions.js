// actions
import apollo from '../../helpers/apolloClient'
import MutationBuilder from '../../helpers/mutationBuilder'
import gql from 'graphql-tag'

const actions = {
  async loadImageCollectionGql (context, resource) {
      let id = resource.id
      console.time(`getResourceById ${id}`)

      const query = gql`
        query GetResource($id: ID!) {
          resource(id: $id) {
             id,
             label,
             viewingHint,
             thumbnail {
               id,
               thumbnailUrl,
               iiifServiceUrl
             },
             ... on ScannedResource {
               viewingDirection,
               startPage
             },
             __typename,
             members {
               id,
               label,
               thumbnail {
                id,
               	thumbnailUrl,
                iiifServiceUrl
               },
               viewingHint,
               __typename
             }
          }
        }`

      const variables = {
        id: id
      }

      try {
        const response = await apollo.query({
          query, variables
        })
        context.commit('SET_RESOURCE', response.data.resource)
      } catch(err) {
        context.commit('CHANGE_MANIFEST_LOAD_STATE', 'LOADING_ERROR')
        console.error(err)
      }

      console.timeEnd(`getResourceById ${resource.id}`)

  },
  async saveStateGql (context, resource) {
    window.resource = resource
    let newResource = resource.body
    let newFilesets = resource.filesets

    let mb = new MutationBuilder(newResource, newFilesets)

    const template = mb.build()
    const mutation = gql`${template}`
    const variables = mb.variables()

    try {
      const response = await apollo.mutate({
        mutation, variables
      })
      // reset the state to reflect applied changes
      context.commit('SAVED_STATE', 'SAVED')
      setTimeout(function(){ context.commit('APPLY_STATE') }, 1000);
    } catch(err) {
      context.commit('SAVED_STATE', 'ERROR')
      console.error(err)
    }
  },
}

export default actions
