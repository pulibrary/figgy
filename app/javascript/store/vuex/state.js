// state
const state = {
  id: '',
  bibid: '',
  resourceClassName: '',
  isMultiVolume: false,
  startPage: '',
  thumbnail: '',
  viewingDirection: '',
  viewingHint: '',
  images: [],
  selected: [],
  ogImages: [],
  changeList: [],
  ogState: {},
  manifestLoadState: 'NOT_LOADED',
  cut: [],
  count: 10
}

export default state
