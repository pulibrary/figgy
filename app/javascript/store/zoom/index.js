export const zoomState = {
  zoomed: null,
}

export const zoomMutations = {
  ZOOM(state, obj) {
    state.zoomed = obj
  },
  RESET_ZOOM(state) {
    state.zoomed = null
  },
}

export const zoomGetters = {
  getIsZoomed: state => {
    return state.zoomed
  },
}
