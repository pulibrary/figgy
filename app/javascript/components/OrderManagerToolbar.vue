<template>
  <component
    :is="type"
    :class="['lux-toolbar']"
  >
    <dropdown-menu
      button-label="Selection Options"
      :menu-items="[
        {name: 'All', component: 'All'},
        {name: 'None', component: 'None'},
        {name: 'Alternate', component: 'Alternate', disabled: true},
        {name: 'Inverse', component: 'Inverse'}
      ]"
      @menu-item-clicked="menuSelection($event)"
    />
    <dropdown-menu
      button-label="With Selected..."
      :menu-items="[
        {name: 'Cut', component: 'Cut', disabled: isCutDisabled()},
        {name: 'Paste Before', component: 'Paste Before', disabled: isPasteDisabled()},
        {name: 'Paste After', component: 'Paste After', disabled: isPasteDisabled()}
      ]"
      @menu-item-clicked="menuSelection($event)"
    />
    <spacer />
    <div class="lux-zoom-slider">
      <lux-icon-base
        class="lux-svg-icon"
        icon-name="shrink"
        icon-color="rgb(0,0,0)"
        width="12"
        height="12"
      >
        <lux-icon-picture />
      </lux-icon-base>
      <label for="img_zoom">
        Image zoom
      </label>
      <input
        id="img_zoom"
        type="range"
        min="40"
        max="500"
        value="300"
        @input="resizeCards($event)"
      >
      <lux-icon-base
        class="lux-svg-icon"
        icon-name="grow"
        icon-color="rgb(0,0,0)"
        width="24"
        height="24"
      >
        <lux-icon-picture />
      </lux-icon-base>
    </div>
  </component>
</template>

<script>
import { mapState } from 'vuex'
/**
 * Toolbars allows a user to select a value from a series of options.
 */
export default {
  name: 'Toolbar',
  status: 'ready',
  release: '1.0.0',
  type: 'Pattern',
  props: {
    /**
     * The html element name used for the container
     */
    type: {
      type: String,
      default: 'div'
    }
  },
  computed: {
    ...mapState({
      resource: state => state.ordermanager.resource,
      gallery: state => state.gallery
    }),
    cut: {
      get () {
        return this.gallery.cut
      }
    }
  },
  methods: {
    cutSelected: function () {
      this.$store.dispatch('cut', this.gallery.selected)
      this.selectNone()
    },
    getItemIndexById: function (id) {
      return this.gallery.items
        .map(function (item) {
          return item.id
        })
        .indexOf(id)
    },
    isCutDisabled: function () {
      return !!this.gallery.cut.length
    },
    isPasteDisabled: function () {
      return !(this.gallery.cut.length && this.gallery.selected.length)
    },
    paste: function (indexModifier) {
      let items = this.gallery.items
      items = items.filter(val => !this.gallery.cut.includes(val))
      let pasteAfterIndex =
        this.getItemIndexById(this.gallery.selected[this.gallery.selected.length - 1].id) + indexModifier
      items.splice(pasteAfterIndex, 0, ...this.gallery.cut)
      this.$store.dispatch('paste', items)
      this.resetCut()
      this.selectNone()
    },
    resetCut: function () {
      this.$store.dispatch('cut', [])
    },
    resizeCards: function (event) {
      this.$emit('cards-resized', event)
    },
    menuSelection (value) {
      switch (value.target.innerText) {
        case 'All':
          this.selectAll()
          break
        case 'None':
          this.selectNone()
          break
        case 'Alternate':
          this.selectAlternate()
          break
        case 'Inverse':
          this.selectInverse()
          break
        case 'Cut':
          this.cutSelected()
          break
        case 'Paste Before':
          this.paste(-1)
          break
        case 'Paste After':
          this.paste(1)
          break
      }
    },
    selectAll: function () {
      this.$store.dispatch('select', this.gallery.items)
    },
    selectAlternate: function () {
      let selected = []
      let itemTotal = this.gallery.items.length
      for (let i = 0; i < itemTotal; i = i + 2) {
        selected.push(this.gallery.items[i])
      }
      this.$store.dispatch('select', selected)
    },
    selectInverse: function () {
      let selected = []
      let itemTotal = this.gallery.items.length
      for (let i = 1; i < itemTotal; i = i + 2) {
        selected.push(this.gallery.items[i])
      }
      this.$store.dispatch('select', selected)
    },
    selectNone: function () {
      this.$store.dispatch('select', [])
    }
  }
}
</script>

<style lang="scss" scoped>
.lux-toolbar {
  box-sizing: border-box;
  margin: 0;
  margin-bottom: 16px;
  font-family: franklin-gothic-urw,Helvetica,Arial,sans-serif;
  font-size: 16px;
  line-height: 1;
  background: #f5f5f5;
  height: 64px;
  align-items: center;
  display: flex;
  padding: 0 24px;
}

.lux-zoom-slider {
  .lux-svg-icon,
  input {
    vertical-align: middle;
    line-height: 1;
    margin: 0;
  }

  input[type="range"] {
    display: inline;
    width: auto;
  }

  label {
    position: absolute;
    clip: rect(1px, 1px, 1px, 1px);
    padding: 0;
    border: 0;
    height: 1px;
    width: 1px;
    overflow: hidden;
  }
}
</style>
