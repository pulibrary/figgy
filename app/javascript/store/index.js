import Vue from 'vue/dist/vue.esm'
import Vuex from 'vuex'
import actions from './vuex/actions'
import {modules} from 'lux-design-system'
Vue.use(Vuex)

const store = new Vuex.Store({
  actions,
  modules: {
    ordermanager: modules.resourceModule,
    gallery: modules.galleryModule,
  }
})

export default store
