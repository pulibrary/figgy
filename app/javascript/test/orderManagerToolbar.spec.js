import { createStore } from "vuex"
import { mount, shallowMount } from "@vue/test-utils"
import OrderManagerToolbar from "../components/OrderManagerToolbar.vue"
import { resourceMutations, resourceGetters } from "../store/resource"
import { galleryModule } from '../store/gallery'

let wrapper
let getters
let options
let store
let actions
let items = [
  { id: "a", title: "A", viewingHint: "single", caption: "a", mediaUrl: "https://picsum.photos/600/300/?random" },
  { id: "b", title: "B", caption: "b", mediaUrl: "https://picsum.photos/600/300/?random" },
  { id: "c", title: "C", caption: "c", mediaUrl: "https://picsum.photos/600/300/?random" },
]

describe("OrderManagerToolbar.vue", () => {
  beforeEach(() => {
    actions = {
      cut: vi.fn(),
      paste: vi.fn(),
      select: vi.fn(),
    }

    const gallery = {
      state: {
        items: items,
        selected: [items[0]],
        cut: [],
        changeList: [],
        ogItems: items,
      },
      mutations: galleryModule.mutations,
    }

    const getters = {
      getMemberCount: () => 3
    }

    const resource = {
      state: {
        resource: {
          id: "",
          resourceClassName: "",
          bibId: "9946093213506421",
          label: "Resource not available.",
          thumbnail: "",
          startCanvas: "",
          isMultiVolume: false,
          viewingHint: "single",
          viewingDirection: null,
          members: [],
          loadState: "NOT_LOADED",
          saveState: "NOT_SAVED",
          ogState: {},
        },
      },
      mutations: resourceMutations,
      getters: getters,
      actions: actions,
      modules: {
        gallery: gallery,
      },
    }

    store = createStore({
      modules: {
        ordermanager: resource,
        gallery: gallery,
      },
    })

    options = {
      computed: {
        viewingDirection: function () {
          return 'bottom-to-top'
        },
        viewingHint: function () {
          return 'continuous'
        }
      }
    }

    wrapper = mount(OrderManagerToolbar, { 
      global: {
        options,
        plugins: [store],
        stubs: ["dropdown-menu","lux-icon-base","lux-icon-picture","spacer"],
      }
    })
  })

  it('selects all thumbnails when Select All button is clicked', () => {
    wrapper.vm.selectAll()
    expect(actions.select).toHaveBeenCalled()
  })

  it('selects alternate thumbnails when Select Alternate button is clicked', () => {
    wrapper.vm.selectAlternate()
    expect(actions.select).toHaveBeenCalled()
  })

  it('selects inverse thumbnails when Select Inverse button is clicked', () => {
    wrapper.vm.selectInverse()
    expect(actions.select).toHaveBeenCalled()
  })

  it('deselects all thumbnails when Select None button is clicked', () => {
    wrapper.vm.selectNone()
    expect(actions.select).toHaveBeenCalled()
  })

  it('cut and paste functions work', () => {
    wrapper.vm.cutSelected()
    expect(actions.cut).toHaveBeenCalled()
    wrapper.vm.paste()
    expect(actions.paste).toHaveBeenCalled()
  })

  it("has the expected html structure", () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
