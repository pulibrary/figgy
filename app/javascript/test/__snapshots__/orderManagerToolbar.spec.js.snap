// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderManagerToolbar.vue has the expected html structure 1`] = `
<div
  class="lux-toolbar"
>
  <dropdown-menu-stub
    button-label="Selection Options"
    menu-items="[object Object],[object Object],[object Object],[object Object]"
  />
   
  <dropdown-menu-stub
    button-label="With Selected..."
    menu-items="[object Object],[object Object],[object Object]"
  />
   
  <spacer-stub />
   
  <div
    class="lux-zoom-slider"
  >
    <lux-icon-base-stub
      class="lux-svg-icon"
      height="12"
      icon-color="rgb(0,0,0)"
      icon-name="shrink"
      width="12"
    >
      <lux-icon-picture-stub />
    </lux-icon-base-stub>
     
    <label
      for="img_zoom"
    >
      
      Image zoom
    
    </label>
     
    <input
      id="img_zoom"
      max="500"
      min="40"
      type="range"
      value="300"
    />
     
    <lux-icon-base-stub
      class="lux-svg-icon"
      height="24"
      icon-color="rgb(0,0,0)"
      icon-name="grow"
      width="24"
    >
      <lux-icon-picture-stub />
    </lux-icon-base-stub>
  </div>
</div>
`;
